tasks/image-segmentation-yolov8/src/validate.py

```diff
@@ -0,0 +1,238 @@
+from concurrent.futures import ProcessPoolExecutor, Future, as_completed
+from pathlib import Path
+
+import logging
+import gc
+import csv
+import multiprocessing as mp
+
+from PIL import Image, ImageDraw
+from coretex import TaskRun, ImageDataset, ImageDatasetClasses, folder_manager, TaskRunStatus
+from ultralytics import YOLO
+from matplotlib import pyplot as plt
+from ultralytics.engine.results import Results
+
+import numpy as np
+import torch
+
+
+def createDatasetResults(taskRun: TaskRun[ImageDataset], sampleResults: list[dict[str, str]]) -> None:
+    datasetResult: list[dict[str, str]] = []
+
+    for label in taskRun.dataset.classes.labels:
+        classAcc = [float(result[label]) for result in sampleResults]
+        datasetResult.append({
+            "Class": label,
+            "Accuracy": f"{np.mean(classAcc):.2f}",
+            "Accuracy stdev": f"{np.std(classAcc):.2f}"
+        })
+
+    datasetAcc = [float(result['accuracy']) for result in sampleResults]
+    datasetResult.append({
+        "Class": "dataset",
+        "Accuracy": f"{np.mean(datasetAcc):.2f}",
+        "Accuracy stdev": f"{np.std(datasetAcc):.2f}"
+    })
+
+    datasetResultPath = folder_manager.temp / "dataset_results.csv"
+    with datasetResultPath.open("w") as file:
+        writer = csv.DictWriter(file, fieldnames = datasetResult[0].keys())
+        writer.writeheader()
+        writer.writerows(datasetResult)
+
+    if taskRun.createArtifact(datasetResultPath, datasetResultPath.name) is None:
+        logging.warning(">> [Image Segmentation] Failed to upload csv file with dataset results as artifact")
+    else:
+        logging.info(">> [Image Segmentation] The csv file with dataset results has been uploaded as artifact")
+
+
+def createSampleResults(taskRun: TaskRun[ImageDataset], sampleResults: list[dict[str, str]]) -> None:
+    csvSamplesPath = folder_manager.temp / "sample_results.csv"
+
+    resultsForCsv: list[dict[str, str]] = []
+    for result in sampleResults:
+        resultsForCsv.append({index: result[index] for index in ["sample id", "sample name", "accuracy"]})
```

Comment:
I would not add this data to a list first, but rather call `writer.writerow` directly for this data